File: backend/src/domain/publishing/domain.rs
#![allow(dead_code)]
use std::sync::Arc;

use crate::{
    datasource::{dbdao::dao::DBDao, scylladao::dao::ScyllaDao, vectordao::dao::VectorDao},
    domain::publishing::schema::{PublishLog, PublishTask, Publisher},
    infra::registry::Registry,
};

#[derive(Clone)]
pub struct PublishingDomain {
    pub db_dao: Arc<DBDao>,
    pub scylla_dao: Arc<ScyllaDao>,
    pub vector_dao: Arc<VectorDao>,
}

impl PublishingDomain {
    pub fn new(r: &Registry) -> Self {
        Self {
            db_dao: r.db_dao.clone(),
            scylla_dao: r.scylla_dao.clone(),
            vector_dao: r.vector_dao.clone(),
        }
    }

    pub async fn get_publisher(&self, id: i64) -> Option<Publisher> {
        self.db_dao
            .get_publisher(id)
            .await
            .ok()
            .map(|record| Publisher {
                id: record.id,
                name: record.name,
                platform: record.platform,
                platform_id: record.platform_id,
                enabled: Some(record.enabled),
                created_at: record.created_at,
                updated_at: None,
            })
    }

    pub async fn search_publishers(
        &self,
        platform: Option<String>,
        enabled: Option<bool>,
        page: i64,
        limit: i64,
    ) -> (Vec<Publisher>, i64) {
        match self
            .db_dao
            .search_publishers(platform, enabled, page, limit)
            .await
        {
            Ok(records) => {
                let publishers: Vec<Publisher> = records
                    .into_iter()
                    .map(|record| Publisher {
                        id: record.id,
                        name: record.name,
                        platform: record.platform,
                        platform_id: record.platform_id,
                        enabled: Some(record.enabled),
                        created_at: record.created_at,
                        updated_at: None,
                    })
                    .collect();
                let total = publishers.len() as i64;
                (publishers, total)
            }
            Err(_) => (Vec::new(), 0),
        }
    }

    pub async fn add_publisher(
        &self,
        name: String,
        platform: String,
        platform_id: Option<String>,
        enabled: bool,
    ) -> Result<i64, String> {
        self.db_dao
            .add_publisher(name.as_str(), platform.as_str(), platform_id, enabled)
            .await
            .map_err(|e| e.to_string())
    }

    pub async fn update_publisher(
        &self,
        id: i64,
        name: Option<String>,
        platform: Option<String>,
        platform_id: Option<String>,
        enabled: Option<bool>,
    ) -> Result<i64, String> {
        self.db_dao
            .update_publisher(id, name, platform, platform_id, enabled)
            .await
            .map_err(|e| e.to_string())
    }

    pub async fn publish_content(&self, publisher_id: i64, content_id: i64) -> Result<i64, String> {
        self.db_dao
            .create_publish_task(publisher_id, content_id)
            .await
            .map_err(|e| e.to_string())
    }

    pub async fn get_publish_task(&self, id: i64) -> Option<PublishTask> {
        self.db_dao
            .get_publish_task(id)
            .await
            .ok()
            .map(|record| PublishTask {
                id: record.id,
                publisher_id: record.publisher_id,
                content_id: record.content_id,
                content: record.content,
                status: record.status,
                error: None,
                created_at: record.created_at,
                started_at: None,
                completed_at: None,
            })
    }

    pub async fn search_publish_tasks(
        &self,
        status: Option<String>,
        platform: Option<String>,
        created_at: Option<String>,
        page: i64,
        limit: i64,
    ) -> (Vec<PublishTask>, i64) {
        match self
            .db_dao
            .search_publish_tasks(status, platform, created_at, page, limit)
            .await
        {
            Ok(records) => {
                let tasks: Vec<PublishTask> = records
                    .into_iter()
                    .map(|record| PublishTask {
                        id: record.id,
                        publisher_id: record.publisher_id,
                        content_id: record.content_id,
                        content: record.content,
                        status: record.status,
                        error: None,
                        created_at: record.created_at,
                        started_at: None,
                        completed_at: None,
                    })
                    .collect();
                let total = tasks.len() as i64;
                (tasks, total)
            }
            Err(_) => (Vec::new(), 0),
        }
    }

    pub async fn get_publish_log(&self, id: i64) -> Option<PublishLog> {
        self.db_dao
            .get_publish_log(id)
            .await
            .ok()
            .map(|record| PublishLog {
                id: record.id,
                publish_task_id: record.publish_task_id,
                log_type: record.log_type,
                message: record.message,
                created_at: record.created_at,
            })
    }

    pub async fn search_publish_logs(
        &self,
        publish_task_id: i64,
        log_type: Option<String>,
        page: i64,
        limit: i64,
    ) -> (Vec<PublishLog>, i64) {
        match self
            .db_dao
            .search_publish_logs(publish_task_id, log_type, page, limit)
            .await
        {
            Ok(records) => {
                let logs: Vec<PublishLog> = records
                    .into_iter()
                    .map(|record| PublishLog {
                        id: record.id,
                        publish_task_id: record.publish_task_id,
                        log_type: record.log_type,
                        message: record.message,
                        created_at: record.created_at,
                    })
                    .collect();
                let total = logs.len() as i64;
                (logs, total)
            }
            Err(_) => (Vec::new(), 0),
        }
    }
}

File: backend/domain/publishing/domain.go
package publishing

import (
	"backend/datasource/dbdao"
	"context"
	"errors"
)

type PublishingDomain struct {
	DB *dbdao.DB
}

func NewPublishingDomain(db *dbdao.DB) *PublishingDomain {
	return &PublishingDomain{DB: db}
}

func (d *PublishingDomain) GetPublisher(ctx context.Context, id int64) (*Publisher, error) {
	p, err := d.DB.GetPublisher(ctx, id)
	if err != nil {
		return nil, err
	}
	return &Publisher{
		ID:         p.ID,
		Name:       p.Name,
		Platform:   p.Platform,
		PlatformID: p.PlatformID,
		Enabled:    p.Enabled,
		CreatedAt:  &p.CreatedAt,
		UpdatedAt:  &p.UpdatedAt,
	}, nil
}

func (d *PublishingDomain) SearchPublishers(ctx context.Context, platform string, enabled *bool, page, limit int64) ([]*Publisher, int64, error) {
	if page < 1 {
		page = 1
	}
	if limit < 1 || limit > 100 {
		limit = 20
	}
	ps, total, err := d.DB.SearchPublishers(ctx, platform, enabled, page, limit)
	if err != nil {
		return nil, 0, err
	}
	result := make([]*Publisher, len(ps))
	for i, p := range ps {
		result[i] = &Publisher{
			ID:         p.ID,
			Name:       p.Name,
			Platform:   p.Platform,
			PlatformID: p.PlatformID,
			Enabled:    p.Enabled,
			CreatedAt:  &p.CreatedAt,
			UpdatedAt:  &p.UpdatedAt,
		}
	}
	return result, total, nil
}

func (d *PublishingDomain) AddPublisher(ctx context.Context, name, platform, platformID string, enabled bool) (*Publisher, error) {
	if name == "" || platform == "" {
		return nil, errors.New("missing required fields")
	}
	p := &dbdao.Publisher{
		Name:       name,
		Platform:   platform,
		PlatformID: platformID,
		Enabled:    enabled,
	}
	created, err := d.DB.CreatePublisher(ctx, p)
	if err != nil {
		return nil, err
	}
	return &Publisher{
		ID:         created.ID,
		Name:       created.Name,
		Platform:   created.Platform,
		PlatformID: created.PlatformID,
		Enabled:    created.Enabled,
		CreatedAt:  &created.CreatedAt,
		UpdatedAt:  &created.UpdatedAt,
	}, nil
}

func (d *PublishingDomain) UpdatePublisher(ctx context.Context, id int64, name, platform, platformID *string, enabled *bool) (*Publisher, error) {
	if id <= 0 {
		return nil, errors.New("invalid id")
	}
	p := &dbdao.Publisher{ID: id}
	if name != nil {
		p.Name = *name
	}
	if platform != nil {
		p.Platform = *platform
	}
	if platformID != nil {
		p.PlatformID = *platformID
	}
	if enabled != nil {
		p.Enabled = *enabled
	}
	updated, err := d.DB.UpdatePublisher(ctx, p)
	if err != nil {
		return nil, err
	}
	return &Publisher{
		ID:         updated.ID,
		Name:       updated.Name,
		Platform:   updated.Platform,
		PlatformID: updated.PlatformID,
		Enabled:    updated.Enabled,
		CreatedAt:  &updated.CreatedAt,
		UpdatedAt:  &updated.UpdatedAt,
	}, nil
}

func (d *PublishingDomain) PublishContent(ctx context.Context, publisherID, contentID int64) (int64, error) {
	if publisherID <= 0 || contentID <= 0 {
		return 0, errors.New("invalid publisher_id or content_id")
	}
	task := &dbdao.PublishTask{
		PublisherID: publisherID,
		ContentID:   contentID,
		Status:      "pending",
	}
	created, err := d.DB.CreatePublishTask(ctx, task)
	if err != nil {
		return 0, err
	}
	return created.ID, nil
}

func (d *PublishingDomain) GetPublishTask(ctx context.Context, id int64) (*PublishTask, error) {
	t, err := d.DB.GetPublishTask(ctx, id)
	if err != nil {
		return nil, err
	}
	return &PublishTask{
		ID:          t.ID,
		PublisherID: t.PublisherID,
		ContentID:   t.ContentID,
		Content:     t.Content,
		Status:      t.Status,
		Error:       t.Error,
		CreatedAt:   &t.CreatedAt,
		StartedAt:   t.StartedAt,
		CompletedAt: t.CompletedAt,
	}, nil
}

func (d *PublishingDomain) SearchPublishTasks(ctx context.Context, status, platform, createdAt string, page, limit int64) ([]*PublishTask, int64, error) {
	if page < 1 {
		page = 1
	}
	if limit < 1 || limit > 100 {
		limit = 20
	}
	ts, total, err := d.DB.SearchPublishTasks(ctx, status, platform, createdAt, page, limit)
	if err != nil {
		return nil, 0, err
	}
	result := make([]*PublishTask, len(ts))
	for i, t := range ts {
		result[i] = &PublishTask{
			ID:          t.ID,
			PublisherID: t.PublisherID,
			ContentID:   t.ContentID,
			Content:     t.Content,
			Status:      t.Status,
			Error:       t.Error,
			CreatedAt:   &t.CreatedAt,
			StartedAt:   t.StartedAt,
			CompletedAt: t.CompletedAt,
		}
	}
	return result, total, nil
}

func (d *PublishingDomain) GetPublishLog(ctx context.Context, id int64) (*PublishLog, error) {
	l, err := d.DB.GetPublishLog(ctx, id)
	if err != nil {
		return nil, err
	}
	return &PublishLog{
		ID:            l.ID,
		PublishTaskID: l.PublishTaskID,
		LogType:       l.LogType,
		Message:       l.Message,
		CreatedAt:     &l.CreatedAt,
	}, nil
}

func (d *PublishingDomain) SearchPublishLogs(ctx context.Context, publishTaskID int64, logType string, page, limit int64) ([]*PublishLog, int64, error) {
	if page < 1 {
		page = 1
	}
	if limit < 1 || limit > 100 {
		limit = 20
	}
	ls, total, err := d.DB.SearchPublishLogs(ctx, publishTaskID, logType, page, limit)
	if err != nil {
		return nil, 0, err
	}
	result := make([]*PublishLog, len(ls))
	for i, l := range ls {
		result[i] = &PublishLog{
			ID:            l.ID,
			PublishTaskID: l.PublishTaskID,
			LogType:       l.LogType,
			Message:       l.Message,
			CreatedAt:     &l.CreatedAt,
		}
	}
	return result, total, nil
}
